torch/csrc/jit/codegen/cuda/ir_interface_nodes.h

```diff
@@ -23,59 +23,59 @@ class ViewTransform;
-
-  bool isSymbolic() const {
-    return !(maybe_value_.has_value());
-  }
-  bool isConst() const final {
-    return maybe_value_.has_value();
-  }
-  c10::optional<bool> value() const {
-    return maybe_value_;
-  }
-
-  bool sameAs(const Statement* other) const override;
-
- private:
-  const c10::optional<bool> maybe_value_;
-};
-
-//! A floating-point value. This value can be a symbolic value (defined after
+//! A scalr value. This value can be a symbolic value (defined after
 //! the kernel is compiled) or a constant value (inlined into the kernel
 //! definition).
-template <DataType DT>
-class TORCH_CUDA_CU_API FloatingPoint : public Val {
+template <typename UnderlyingType>
+class TORCH_CUDA_CU_API Scalar : public Val {
  public:
-  using ScalarType = typename DataTypeToNativeType<DT>::type;
+  using ScalarType = UnderlyingType;
+  static constexpr DataType kDefaultDataType =
+      NativeTypeToDataType<UnderlyingType>::type;
 
-  FloatingPoint(IrBuilderPasskey passkey)
-      : Val(passkey, ValType::Scalar, DT), maybe_value_{c10::nullopt} {}
+  explicit Scalar(IrBuilderPasskey passkey, DataType dtype = kDefaultDataType)
+      : Val(passkey, ValType::Scalar, dtype), maybe_value_{c10::nullopt} {
+    TORCH_INTERNAL_ASSERT(
+        (std::is_integral<UnderlyingType>::value && isIntegralType(dtype)) ||
+            (std::is_same<UnderlyingType, bool>::value &&
+             isBooleanType(dtype)) ||
+            (std::is_floating_point<UnderlyingType>::value &&
+             isFloatingPointType(dtype)),
+        "Invalid data type: ",
+        dtype);
+  }
 
-  explicit FloatingPoint(IrBuilderPasskey passkey, ScalarType value)
-      : Val(passkey, ValType::Scalar, DT), maybe_value_{value} {}
+  explicit Scalar(
+      IrBuilderPasskey passkey,
+      UnderlyingType value,
+      DataType dtype = kDefaultDataType)
+      : Val(passkey, ValType::Scalar, dtype), maybe_value_{value} {
+    TORCH_INTERNAL_ASSERT(
+        (std::is_integral<UnderlyingType>::value && isIntegralType(dtype)) ||
+            (std::is_same<UnderlyingType, bool>::value &&
+             isBooleanType(dtype)) ||
+            (std::is_floating_point<UnderlyingType>::value &&
+             isFloatingPointType(dtype)),
+        "Invalid data type: ",
+        dtype);
+  }
 
-  explicit FloatingPoint(
+  explicit Scalar(
       IrBuilderPasskey passkey,
-      c10::optional<ScalarType> value)
-      : Val(passkey, ValType::Scalar, DT), maybe_value_{value} {}
+      c10::optional<UnderlyingType> value,
+      DataType dtype = kDefaultDataType)
+      : Val(passkey, ValType::Scalar, dtype), maybe_value_{value} {
+    TORCH_INTERNAL_ASSERT(
+        (std::is_integral<UnderlyingType>::value && isIntegralType(dtype)) ||
+            (std::is_same<UnderlyingType, bool>::value &&
+             isBooleanType(dtype)) ||
+            (std::is_floating_point<UnderlyingType>::value &&
+             isFloatingPointType(dtype)),
+        "Invalid data type: ",
+        dtype);
+  }
```

Comment:
Could these two ctors be merge into a single one? I think only keeping the `c10::optional` version should be fine, and `UnderlyingType` could be automatically converted to `c10::optional<UnderlyingType>`